fairseq/sequence_generator.py

```diff
@@ -367,6 +381,14 @@ def _generate(
             if self.no_repeat_ngram_size > 0:
                 lprobs = self._no_repeat_ngram(tokens, lprobs, bsz, beam_size, step)
 
+            # before step action, set the probabilities of negative constraints to -inf
+            negative_indices = self.search.get_negative_tokens(negative_active_hypos)
+            if negative_indices is not None:
+                negative_indices = torch.tensor(negative_indices, device=lprobs.device)
+                if negative_indices.numel():
+                    inf_value = torch.ones(negative_indices.shape[0], device=lprobs.device)*-math.inf
+                    lprobs.index_put_(tuple(negative_indices.t()), inf_value)
```

Comment:
This block can move to the beginning of `LexicallyConstrainedBeamSearch::step`. The only thing it depends on is `lprobs`, which is already passed to step, and `negative_active_hypos`, which can be maintained internally by `LexicallyConstrainedBeamSearch` (see comment below).

Reply:
Done.